vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -184,40 +225,29 @@ void remove_slot_features(example* shared, example* slot)
   }
 }
 
-// Generates all combinations of 4th order interactions between namespaces in [shared,history,slot,action]
+// Generates interaction between all namespaces and interactions with the slot id namespace.
 void calculate_and_insert_interactions(example* shared, example* slot, std::vector<example*> actions, std::vector<std::string>& vec)
 {
-  vec.reserve(shared->indices.size() * slot->indices.size() + vec.size());
+  std::vector<std::string> new_interactions;
+  for(auto interaction : vec)
+  {
+    interaction.push_back((char)ccb_id_namespace);
+    new_interactions.push_back(interaction);
+  }
+  vec.insert(vec.end(), new_interactions.begin(), new_interactions.end());
 
-  for (auto shared_index : shared->indices)
+  for (auto& action : actions)
   {
-    for (auto action : actions)
+    for (auto& action_index : action->indices)
```

Comment:
I'm structurally confused by the function.  It seems like you should only map slot features into shared features?  Why all the extra work?

Reply:
This function is determining all interactions that need to occur. Moving the slot features to the shared example is done with `inject_slot_features`. This specific block is creating the quadratics for actions.

Reply:
Ah, good, I understand now.